File: src/_arch/aarch64/cpu.zig
pub fn spin_for_cycles(n: usize) void {
    for (0..n) |_| {
        asm volatile ("nop");
    }
}

pub fn wait_forever() noreturn {
    while (true) {
        asm volatile ("wfe");
    }
}

pub fn nop() void {
    asm volatile ("nop");
}

File: src/kernel/arch/aarch64/cpu.cpp
#include "../cpu.hpp"

namespace CPU {
size_t currentCoreId() {
  unsigned long mpidr = 0;
  asm volatile("mrs %0, MPIDR_EL1" : "=r"(mpidr));
  return static_cast<size_t>(mpidr & 0b11);
}

void spinForCycles(unsigned int cycles) {
  for (unsigned int i = 0; i < cycles; i++) {
    asm volatile("nop");
  }
}

void nop() { asm volatile("nop"); }

void waitForEvent() { asm volatile("wfe" ::: "memory"); }

void sendEvent() { asm volatile("sev" ::: "memory"); }

void waitForever() {
  while (true) {
    waitForEvent();
  }
}

} // namespace CPU
